crates/xilem_masonry/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+        self.pod.on_pointer_event(ctx, event);
+    }
+    fn on_text_event(&mut self, ctx: &mut EventCtx, event: &TextEvent) {
+        self.pod.on_text_event(ctx, event);
+    }
+
+    fn on_status_change(&mut self, _: &mut LifeCycleCtx, _: &StatusChange) {
+        // Intentionally do nothing?
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle) {
+        self.pod.lifecycle(ctx, event);
+    }
+
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
+        let size = self.pod.layout(ctx, bc);
+        ctx.place_child(&mut self.pod, Point::ORIGIN);
+        size
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
+        self.pod.paint(ctx, scene);
+    }
+
+    fn children(&self) -> SmallVec<[WidgetRef<'_, dyn Widget>; 16]> {
+        let mut vec = SmallVec::new();
+        vec.push(self.pod.as_dyn());
+        vec
+    }
+}
+
+impl<State, Logic, View> AppDriver for MasonryDriver<State, Logic, View>
+where
+    Logic: FnMut(&mut State) -> View,
+    View: MasonryView<State>,
+{
+    fn on_action(
+        &mut self,
+        ctx: &mut masonry::app_driver::DriverCtx<'_>,
+        widget_id: masonry::WidgetId,
+        action: masonry::Action,
+    ) {
+        if let Some(id_path) = self.view_cx.widget_map.get(&widget_id) {
+            let message_result =
+                self.current_view
+                    .message(id_path.as_slice(), Box::new(action), &mut self.state);
+            let rebuild = match message_result {
+                MessageResult::Action(()) => {
+                    // It's not entirely clear what to do here
+                    true
+                }
+                MessageResult::RequestRebuild => true,
+                MessageResult::Nop => false,
+                MessageResult::Stale(_) => {
+                    tracing::info!("Discarding message");
+                    false
+                }
+            };
+            if rebuild {
+                let next_view = (self.logic)(&mut self.state);
+                let mut root = ctx.get_root::<RootWidget<View::Element>>();
+                let element = root.get_element();
+
+                let changed = next_view.rebuild(&mut self.view_cx, &self.current_view, element);
+                if !changed.changed {
+                    // Masonry manages all of this itself - ChangeFlags is probably not needed?
+                    tracing::debug!("TODO: Skip some work?");
+                }
+                self.current_view = next_view;
+            }
+        } else {
+            eprintln!("Got action {action:?} for unknown widget. Did you forget to use `with_action_widget`?");
+        }
+    }
+}
+
+impl<E: Widget + StoreInWidgetMut> RootWidgetMut<'_, E> {
+    pub fn get_element(&mut self) -> WidgetMut<'_, E> {
+        self.ctx.get_mut(&mut self.widget.pod)
+    }
+}
+
+impl<State, Logic, View> Xilem<State, Logic, View>
```

Comment:
This might be too cute, but I'd like to keep it like this

I wanted to avoid the term `App`, because this is plausibly embeddable.